design/ozone.md

```diff
@@ -0,0 +1,246 @@
+
+At the end of the transformations we will have a list of real K8s resources object, where we can do any final adjustment on the API object and send them to the k8s API services.
+
+The main logic is the following: 
+
+ * The resource generator can be initialized with the main k8s definition directory (which is included in the Ozone release, and included in the rook operator image)
+ * The transformations can be added programatically based on CRD properties
+ * The real k8s golang objects are generated
+ * After final adjustment in golang, we can post all the objects to the K8S api
+
+### CRD
+
+A good operator CRD provides a good balance between the simplicity and power. By default the CRD should be as simple as possible but we should keep the possibility to customize any segment of the generated cluster.
+
+This can be achieved by the use k8s generator tool.
+
+ 1. The default cluster will be generated based on the main attributes of CRD
+ 2. Additional attributes can trigger additional transformations
+ 3. Finally we can provide an option to define any custom transformations
+
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-ozone
+---
+apiVersion: ozone.rook.io/v1alpha1
+kind: ObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ozone
+spec:
+  components: #high lebel options to turn off/on component groups
+     csi: true
+     recon: true
+     s3g: true
+  features: # these are not components but confiuration groups
+     prometheus: true
+     tracing: true
+  ozoneVersion: #see ceph versioning doc for mor details
+    image: apache/ozone:0.4.1
+    allowUnsupported: false
+  upgradePolicy: #see ceph versioning doc for mor details (Won't be implmented in Phase 1)
+    ozoneVersion:
+      image: apache/ozone:0.5.0
+      allowUnsupported: false
+  storage: #more storage option should be added to here
+     type: ephemeral/auto/manual # ephemeral: use emptyDir, auto: using the existing rook method to identify local disk, manual: should be specified
+     persistence:
+        metadata: #specify persistent volumes for metadata disk
+        storage: #specify persistent volumes for storage
+  annotations: #custom annotations for all the created k8s resources
+     key: value
+  customization: #deep water: any additional transformations to modify k8s resources. Only for advanced users.
+    - type: Add
+      path: ["metadata", "labels"]
+      value:
+         generated-with: rook
+    
+```
+
+## Operator capabilities
+
+The main benefit to use an operator is to provide additional functionalities based on the current state of the runtime environment. The power of the operator based on the runtime information which is not available for any other installer tool.
+
+Some option which can be implemented as part of the Operator:
+
+ * If only one node is available (eg. Minikube) in the cluster, use `StatefulSet` with `replicas: 3` to make it possible to test full replicatoin. 
```

Comment:
The setting for number of replicas may be more useful in the CR. If the operator makes too many assumptions about the desired state, the user has a hard time understanding. Users will understand the desired state they set in the yaml.

Reply:
Ok. I removed this part.